crypto/src/lib.rs

```diff
@@ -224,21 +230,42 @@ impl KeyPair {
         })
     }
 
-    /// Generates a pair of Public and Private key with the corresponding [`KeyGenConfiguration`].
-    pub fn generate_with_configuration(
-        KeyGenConfiguration {
-            key_gen_option,
-            algorithm,
-        }: KeyGenConfiguration,
-    ) -> Self {
-        match algorithm {
-            Algorithm::Ed25519 => signature::ed25519::Ed25519Sha512::keypair(key_gen_option).into(),
-            Algorithm::Secp256k1 => {
-                signature::secp256k1::EcdsaSecp256k1Sha256::keypair(key_gen_option).into()
-            }
-            Algorithm::BlsNormal => signature::bls::BlsNormal::keypair(key_gen_option).into(),
-            Algorithm::BlsSmall => signature::bls::BlsSmall::keypair(key_gen_option).into(),
+    /// Generates a pair of Public and Private key with the corresponding [`KeyGenOption`].
+    pub fn generate_with_config(config: KeyGenConfig) -> Self {
```

Comment:
instead of having `KeyGenConfig` I think it would be preferable to have 3 methods on `KeyPair`:
```rs
pub fn rand() -> Self;
pub fn from_seed(algorithm: Algorithm, seed: Vec<u8>) -> Self;
impl TryFrom<PrivateKey> for KeyPair {
    //..
}
```

or similar naming

what do you think?

Reply:
Yup, I think it would work totally fine and simplify things:

- `KeyPair::generate_from_random(algorithm)`
- `KeyPair::generate_from_seed(seed, algorithm)`
- `KeyPair::generate_from_private_key(private_key)`

Let me impl it tomorrow.

Reply:
And yeah, `TryFrom<PrivateKey>` makes sense too, although I still having a method for it is not bad.

Also, I think it would make more sense to use **"derive"** term:

- generate from random (or just "generate")
- derive from seed
- derive from private key

Reply:
> And yeah, TryFrom<PrivateKey> makes sense too, although still having a method for it is not bad.

I wouldn't be so eager to increase API surface. This can always be done so, as a compromise, I would make this into a separate issue to discuss and atm just leave `TryFrom`

> Also, I think it would make more sense to use "derive" term:

"derive" doesn't sound quite right to me. I'd vote for either:
1.
```rs
Keypair::random(algorithm: Algorithm)
Keypair::from_seed(algorithm: Algorithm, seed: Vec<u8>) -> Self;
```
2.
```rs
KeyPair::generate_from_random(algorithm: Algorithm)
KeyPair::generate_from_seed(algorithm: Algorithm, seed: Vec<u8>)
```

but I don't find `generate` prefix brings anything but a longer method name

Reply:
AFAIK, _derivation_ is an established term in cryptography. It applies to both deriving a key pair using pRNG + seed, and to deriving of the public key from a private one.

Some references:

- https://en.wikipedia.org/wiki/Key_derivation_function
- https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/deriveKey

Anyway, `KeyPair::random` and `KeyPair::from_seed` works fine, imo. And will leave `TryFrom` as the only way for `PrivateKey`.